framework/include/kernels/ADDiffusion.h

```diff
@@ -0,0 +1,38 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef ADDIFFUSION_H
+#define ADDIFFUSION_H
+
+#include "ADKernel.h"
+
+template <ComputeStage compute_stage>
+class ADDiffusion;
+
+template <>
+InputParameters validParams<ADDiffusion<RESIDUAL>>();
+template <>
+InputParameters validParams<ADDiffusion<JACOBIAN>>();
+
+template <ComputeStage compute_stage>
+class ADDiffusion : public ADKernel<compute_stage>
+{
+public:
+  ADDiffusion(const InputParameters & parameters);
+
+protected:
+  virtual typename ResidualReturnType<compute_stage>::type computeQpResidual() override;
```

Comment:
Thinking of a macro for the return type: `ADResidual` or something?

Reply:
I would love to make the return type here `ADReal`... is that possible?

Also: thinking about these return types... is it a problem that they're returned by copy?  Do copies actually get made or is copy elision done here?